google_compute_engine/accounts/oslogin_utils.py

```diff
@@ -103,37 +103,50 @@ def _RemoveOsLoginNssCache(self):
         if e.errno != errno.ENOENT:
           raise
 
-  def UpdateOsLogin(self, enable_oslogin, two_factor=False):
+  def UpdateOsLogin(self, oslogin_desired, two_factor_desired=False):
     """Update whether OS Login is enabled and update NSS cache if necessary.
 
     Args:
-      enable_oslogin: bool, enable OS Login if True, disable if False.
+      oslogin_desired: bool, enable OS Login if True, disable if False.
       duration: int, number of seconds before updating the NSS cache.
 
     Returns:
       int, the return code from updating OS Login, or None if not present.
     """
-    # Two factor can only be enabled when OS Login is enabled.
-    two_factor = two_factor and enable_oslogin
-    status = self._GetStatus(two_factor=two_factor)
-    if status is None:
+    oslogin_configured = self._GetStatus(two_factor=False)
+    if oslogin_configured is None:
       return None
-
-    if enable_oslogin:
-      if not status:
+    two_factor_configured = self._GetStatus(two_factor=True)
+    # Two factor can only be enabled when OS Login is enabled.
+    two_factor_desired = two_factor_desired and oslogin_desired
+
+    if oslogin_desired:
+      params = ['activate']
+      if two_factor_desired:
+        params += ['--twofactor']
+      # OS Login is desired and not enabled.
+      if not oslogin_configured:
         self.logger.info('Activating OS Login.')
-        params = ['activate']
-        if two_factor:
-          params += ['--twofactor']
         return self._RunOsLoginControl(params) or self._RunOsLoginNssCache()
-      # OS Login is already enabled. Update the cache if appropriate.
+      # Enable two factor authentication.
+      if not two_factor_configured and two_factor_desired:
+        self.logger.info('Activating OS Login with two factor enabled.')
+        return self._RunOsLoginControl(params) or self._RunOsLoginNssCache()
```

Comment:
this `or` is deeply confusing and only works because we pass the unix return code from `_RunOsLoginControl()` which parses as false. explicitly running these functions would be much easier to parse

Reply:
Short circuiting when a command fails is preferred so I'll leave it as is.